File: Python/SelectionSort.py
#coding: utf-8
from Sort import Sort

__author__ = "Ionesio Junior" 

class SelectionSort(Sort):
	
	''' Selection sort class have many ways to implement selection sort idea '''

	def sort(self,array,leftIndex = 0 , rightIndex = None):
        	'''Generic Method
                algorithm.sort(array) -> return array complete ordered
                algorithm.sort(array,leftIndex,rightIndex) -> return array partially ordered
        	'''
		if(rightIndex == None):
			rightIndex = len(array) -1
		if self.__validateParams(array,leftIndex,rightIndex):
			#self.__simpleSelectionSort(array,leftIndex,rightIndex)
			#self.__bidirectionalSelectionSort(array,leftIndex,rightIndex)
			#self.__recursiveSimpleSelectionSort(array,leftIndex,rightIndex)
			self.__recursiveBidirectionalSelectionSort(array,leftIndex,rightIndex)
	
	
	def __simpleSelectionSort(self,array,leftIndex,rightIndex):
        	''' Simple iteractive selection sort implementation '''
		while(leftIndex < rightIndex):
			minIndex = leftIndex;
			for i in range(leftIndex+1,rightIndex+1):
				if(array[minIndex]> array[i]):
					minIndex = i
			array[minIndex],array[leftIndex] = array[leftIndex] ,array[minIndex]
			leftIndex = leftIndex + 1
	

	def __bidirectionalSelectionSort(self,array,leftIndex,rightIndex):
        	''' Bidirectional iteractive selection sort implementation '''		
		while(leftIndex < rightIndex):
			minIndex = leftIndex
			for i in range(leftIndex + 1 ,rightIndex +1):
				if(array[minIndex] > array[i]):
					minIndex = i		
			array[minIndex], array[leftIndex] = array[leftIndex],array[minIndex]
			leftIndex = leftIndex +1
			
			maxIndex = rightIndex
			for i in range(rightIndex-1 , leftIndex-1,-1):
				if(array[maxIndex] < array[i]):
					maxIndex = i
			array[maxIndex] , array[rightIndex] = array[rightIndex],array[maxIndex]
			rightIndex = rightIndex -1

	
	def __recursiveSimpleSelectionSort(self,array,leftIndex,rightIndex):
        	''' Recursive simple  selection sort implementation '''	
		if(leftIndex < rightIndex):
			self.__recursiveGetMin(array,leftIndex,rightIndex,leftIndex,rightIndex)
			self.__recursiveSimpleSelectionSort(array,leftIndex + 1,rightIndex)

	
	def __recursiveBidirectionalSelectionSort(self,array,leftIndex,rightIndex):
		''' Recursive bidirectional selection sort implementation '''	
		if(leftIndex < rightIndex):
			self.__recursiveGetMax(array,leftIndex,rightIndex,rightIndex,rightIndex)
			self.__recursiveGetMin(array,leftIndex,rightIndex,leftIndex,leftIndex)
			self.__recursiveBidirectionalSelectionSort(array,leftIndex + 1,rightIndex-1)

	
	def __recursiveGetMax(self,array,leftIndex,rightIndex,maxValue,index):
		''' This method get max element in array and switch with rightIndex position '''
		if(index >= leftIndex):
			if(array[maxValue] < array[index]):
				maxValue = index
			self.__recursiveGetMax(array,leftIndex,rightIndex,maxValue,index-1)
		else:
			array[maxValue],array[rightIndex] = array[rightIndex] , array[maxValue]


	def __recursiveGetMin(self,array,leftIndex,rightIndex,minValue,index):
                ''' This method get min element in array and switch with leftIndex position '''	
		if(index <= rightIndex):
			if(array[minValue] > array[index]):
				minValue = index
			self.__recursiveGetMin(array,leftIndex,rightIndex,minValue,index +1)
		else:
			array[minValue] ,array[leftIndex] = array[leftIndex] , array[minValue]

	
	def __validateParams(self,array,leftIndex,rightIndex):
                ''' This method verify if parameters are valid if some parameter is not valid the method returns false '''
		if(array == None or leftIndex < 0 or rightIndex > len(array) -1 or leftIndex >= rightIndex):
			return False
		else:
			return True;

File: Java/src/SelectionSort/SelectionSort.java
package SelectionSort;

import sorting.AbstractSort;

public class SelectionSort<T extends Comparable<T>> extends AbstractSort<T> {

	@Override
	public void sort(T[] array, int leftIndex, int rightIndex) {
		if(validateParams(array, leftIndex, rightIndex)){
			//simpleSelectionSort(array, leftIndex, rightIndex);
			//bidirectionalSelectionSort(array, leftIndex, rightIndex);
			recursiveBidirectionalSelectionSort(array, leftIndex, rightIndex);
			//recursiveSimpleSelectionSort(array, leftIndex, rightIndex);
		}else{
			//Do nothing
		}
	}
	
	private void simpleSelectionSort(T[] array,int leftIndex,int rightIndex){
		while(leftIndex < rightIndex){
			int minIndex = leftIndex;
			for(int i = leftIndex+1;i <= rightIndex;i++){
				if(array[minIndex].compareTo(array[i]) > 0){
					minIndex = i;
				}
			}
			swap(array,minIndex,leftIndex++);
		}
	}
	
	private void bidirectionalSelectionSort(T[] array,int leftIndex,int rightIndex){
		while(leftIndex < rightIndex){
			int minIndex = leftIndex;
			for(int i = leftIndex  +1;i <= rightIndex;i++){
				if(array[minIndex].compareTo(array[i]) > 0){
					minIndex = i;
				}
			}
			swap(array,minIndex,leftIndex++);
			
			int maxIndex = rightIndex;
			for(int i = rightIndex;i >= leftIndex;i--){
				if(array[maxIndex].compareTo(array[i]) < 0){
					maxIndex = i;
				}
			}
			swap(array,maxIndex,rightIndex--);
		}
	}

	
	//Recursive Methods
	private void recursiveSimpleSelectionSort(T[] array,int leftIndex,int rightIndex){
		if(leftIndex < rightIndex){
			recursiveGetMin(array, leftIndex, rightIndex, leftIndex, leftIndex);
			recursiveSimpleSelectionSort(array, leftIndex +1, rightIndex);
		}
	}
	
	private void recursiveBidirectionalSelectionSort(T[] array,int leftIndex,int rightIndex){
		if(leftIndex < rightIndex){
			recursiveGetMax(array, leftIndex, rightIndex, rightIndex, rightIndex);
			recursiveGetMin(array, leftIndex, rightIndex, leftIndex, leftIndex);
			recursiveBidirectionalSelectionSort(array, leftIndex +1, rightIndex -1);
		}
	}
	
	
	
	//Recursive Auxiliar Methods
	private void recursiveGetMax(T[] array,int leftIndex,int rightIndex,int max,int index){
		if(index >= leftIndex){
			if(array[max].compareTo(array[index]) < 0){
				max = index;
			}
			recursiveGetMax(array,leftIndex,rightIndex,max,index-1);
		}else{
			swap(array,max,rightIndex);
		}
	}
	
	private void recursiveGetMin(T[] array,int leftIndex,int rightIndex,int min,int index){
		if(index <= rightIndex){
			if(array[min].compareTo(array[index]) > 0){
				min = index;
			}
			recursiveGetMin(array,leftIndex,rightIndex,min,index+1);
		}else{
			swap(array,min,leftIndex);
		}
	}

	
	//Auxiliar Methods
	private void swap(T[] array,int i,int j){
		T aux = array[i];
		array[i] = array[j];
		array[j] = aux;
	}
	
	private boolean validateParams(T[] array,int leftIndex,int rightIndex){
		if(array == null || leftIndex < 0 || leftIndex >= rightIndex || rightIndex > array.length - 1){
			return false;
		}else{
			return true;
		}
	}

}
